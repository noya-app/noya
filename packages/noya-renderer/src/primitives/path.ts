import Sketch from '@sketch-hq/sketch-file-format-ts';
import { CanvasKit, Path, PathOp } from 'canvaskit';
import { distance, Rect } from 'noya-geometry';
import { Point } from 'noya-state';
import { windowsOf } from 'noya-utils';
import { parsePoint, stringifyPoint } from '../primitives';

/**
 * The radius of an edge should be less than half the length of that edge
 */
function getEdgeRadius(a: Point, b: Point, fixedRadius: number): number {
  const dist = distance(a, b);

  return dist <= fixedRadius * 2 ? dist / 2 : fixedRadius;
}

/**
 * The radius of a corner should be symmetric on both sides
 */
function getCornerRadius(
  before: Point,
  corner: Point,
  after: Point,
  fixedRadius: number,
): number {
  const beforeRadius = getEdgeRadius(corner, before, fixedRadius);
  const afterRadius = getEdgeRadius(corner, after, fixedRadius);
  return Math.min(beforeRadius, afterRadius);
}

// Implementation loosely follows:
// https://github.com/google/skia/blob/13d6c4f55db0d1429f889704ed9b2e48968ac7a3/src/effects/SkCornerPathEffect.cpp#L24
function getStep(
  a: Point,
  b: Point,
  radius: number,
): { step: Point; drawSegment: boolean } {
  let dist = distance(a, b);

  const step: Point = { x: b.x - a.x, y: b.y - a.y };

  // Since we clamp radius beforehand, the dist will never be < radius * 2, only equal
  if (dist <= radius * 2) {
    step.x /= 2;
    step.y /= 2;

    return { step, drawSegment: false };
  }

  step.x *= radius / dist;
  step.y *= radius / dist;

  return { step, drawSegment: true };
}

export type ParsedCurvePoint = Omit<
  Sketch.CurvePoint,
  'curveFrom' | 'curveTo' | 'point'
> & {
  curveFrom: Point;
  curveTo: Point;
  point: Point;
};

export function scalePoint(point: Point, { x, y, width, height }: Rect): Point {
  return { x: x + point.x * width, y: y + point.y * height };
}

export function unscalePoint(
  point: Point,
  { x, y, width, height }: Rect,
): Point {
  return { x: (point.x - x) / width, y: (point.y - y) / height };
}

export function parseCurvePoint(
  curvePoint: Sketch.CurvePoint,
  frame: Rect,
): ParsedCurvePoint {
  return {
    ...curvePoint,
    curveFrom: scalePoint(parsePoint(curvePoint.curveTo), frame),
    curveTo: scalePoint(parsePoint(curvePoint.curveFrom), frame),
    point: scalePoint(parsePoint(curvePoint.point), frame),
  };
}

export function unparseCurvePoint(
  curvePoint: ParsedCurvePoint,
  frame: Rect,
): Sketch.CurvePoint {
  return {
    ...curvePoint,
    curveFrom: stringifyPoint(unscalePoint(curvePoint.curveTo, frame)),
    curveTo: stringifyPoint(unscalePoint(curvePoint.curveFrom, frame)),
    point: stringifyPoint(unscalePoint(curvePoint.point, frame)),
  };
}

export function path(
  CanvasKit: CanvasKit,
  points: Sketch.CurvePoint[],
  frame: Rect,
): Path {
  const curvePoints = points.map((point) => parseCurvePoint(point, frame));

  const pairs = windowsOf(curvePoints, 3, true);

  const path = new CanvasKit.Path();

  // Move to the starting point
  if (pairs[pairs.length - 1]) {
    const [prev, current, next] = pairs[pairs.length - 1];
    const currentPoint = current.point;
    const nextPoint = next.point;
    const prevPoint = prev.point;

    if (next.curveMode === Sketch.CurveMode.Straight) {
      const radius = getCornerRadius(
        prevPoint,
        currentPoint,
        nextPoint,
        current.cornerRadius,
      );

      const { step } = getStep(currentPoint, nextPoint, radius);

      path.moveTo(currentPoint.x + step.x, currentPoint.y + step.y);
    } else {
      path.moveTo(nextPoint.x, nextPoint.y);
    }
  }

  pairs.forEach((pair) => {
    const [prev, current, next] = pair;

    const currentPoint = current.point;
    const currentCurveTo = current.curveTo;
    const nextCurveFrom = next.curveFrom;
    const nextPoint = next.point;
    const prevPoint = prev.point;

    if (next.curveMode === Sketch.CurveMode.Straight) {
      const radius = getCornerRadius(
        prevPoint,
        currentPoint,
        nextPoint,
        current.cornerRadius,
      );

      const { step, drawSegment } = getStep(currentPoint, nextPoint, radius);

      if (current.cornerRadius > 0) {
        path.arcToTangent(
          currentPoint.x,
          currentPoint.y,
          currentPoint.x + step.x,
          currentPoint.y + step.y,
          radius,
        );
      } else {
        path.lineTo(currentPoint.x + step.x, currentPoint.y + step.y);
      }

      if (drawSegment) {
        path.lineTo(nextPoint.x - step.x, nextPoint.y - step.y);
      }
    } else {
      path.cubicTo(
        currentCurveTo.x,
        currentCurveTo.y,
        nextCurveFrom.x,
        nextCurveFrom.y,
        nextPoint.x,
        nextPoint.y,
      );
    }
  });

  path.close();

  // console.log(path.toCmds());

  return path;
}

const getBorderPositionPathOp = (
  CanvasKit: CanvasKit,
  position: Sketch.BorderPosition,
): PathOp | undefined => {
  switch (position) {
    case Sketch.BorderPosition.Center:
      return;
    case Sketch.BorderPosition.Outside:
      return CanvasKit.PathOp.Difference;
    case Sketch.BorderPosition.Inside:
      return CanvasKit.PathOp.Intersect;
  }
};

export function getStrokedPath(
  CanvasKit: CanvasKit,
  path: Path,
  width: number,
  pathOp?: PathOp,
) {
  const copy = path.copy();

  // A 0-width path will never change the original when unioned
  if (width === 0 && pathOp === CanvasKit.PathOp.Union) return copy;

  if (
    !copy.stroke({
      width,
    })
  ) {
    console.info('[getStrokedPath] Failed to stroke path');
    return copy;
  }

  if (pathOp !== undefined) {
    if (!copy.op(path, pathOp)) {
      console.info('[getStrokedPath] Failed to combine paths');
    }
  }

  return copy;
}

export function getStrokedBorderPath(
  CanvasKit: CanvasKit,
  path: Path,
  borderWidth: number,
  borderPosition: Sketch.BorderPosition,
) {
  switch (borderPosition) {
    case Sketch.BorderPosition.Center:
      break;
    case Sketch.BorderPosition.Outside:
    case Sketch.BorderPosition.Inside:
      borderWidth *= 2;
  }

  const pathOp = getBorderPositionPathOp(CanvasKit, borderPosition);

  return getStrokedPath(CanvasKit, path, borderWidth, pathOp);
}
