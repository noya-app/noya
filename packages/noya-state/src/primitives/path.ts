import { Sketch } from '@noya-app/noya-file-format';
import { Point, Rect, distance } from '@noya-app/noya-geometry';
import { PointString, SketchModel } from '@noya-app/noya-sketch-model';
import { clamp, rotate, windowsOf, zip } from '@noya-app/noya-utils';
import { CanvasKit, Path, PathOp } from 'canvaskit';
import {
  CommandWithoutQuadratics,
  makePathsFromCommands,
} from 'noya-import-svg';
import {
  lineCapStyle,
  lineJoinStyle,
  parsePoint,
  stringifyPoint,
} from 'noya-state';
import { PathCommand, PathCommandVerb, parsePathCmds } from './pathCommand';

/**
 * The radius of an edge should be less than half the length of that edge
 */
function getEdgeRadius(a: Point, b: Point, cornerRadius: number): number {
  const dist = distance(a, b);

  return dist <= cornerRadius * 2 ? dist / 2 : cornerRadius;
}

/**
 * The radius of a corner should be symmetric on both sides
 */
function getCornerRadius(
  before: Point,
  corner: Point,
  after: Point,
  cornerRadius: number,
): number {
  const beforeRadius = getEdgeRadius(corner, before, cornerRadius);
  const afterRadius = getEdgeRadius(corner, after, cornerRadius);
  return Math.min(beforeRadius, afterRadius);
}

// Implementation loosely follows:
// https://github.com/google/skia/blob/13d6c4f55db0d1429f889704ed9b2e48968ac7a3/src/effects/SkCornerPathEffect.cpp#L24
function getStep(
  a: Point,
  b: Point,
  radius: number,
): { step: Point; drawSegment: boolean } {
  let dist = distance(a, b);

  const step: Point = { x: b.x - a.x, y: b.y - a.y };

  // Since we clamp radius beforehand, the dist will never be < radius * 2, only equal
  if (dist <= radius * 2) {
    step.x /= 2;
    step.y /= 2;

    return { step, drawSegment: false };
  }

  step.x *= radius / dist;
  step.y *= radius / dist;

  return { step, drawSegment: true };
}

export type DecodedCurvePoint = Omit<
  Sketch.CurvePoint,
  'curveFrom' | 'curveTo' | 'point'
> & {
  curveFrom: Point;
  curveTo: Point;
  point: Point;
};

export function scalePoint(point: Point, { x, y, width, height }: Rect): Point {
  return { x: x + point.x * width, y: y + point.y * height };
}

export function unscalePoint(
  point: Point,
  { x, y, width, height }: Rect,
): Point {
  return { x: (point.x - x) / width, y: (point.y - y) / height };
}

export function decodeCurvePoint(
  curvePoint: Sketch.CurvePoint,
  frame: Rect,
): DecodedCurvePoint {
  return {
    ...curvePoint,
    hasCurveFrom: curvePoint.hasCurveTo,
    hasCurveTo: curvePoint.hasCurveFrom,
    curveFrom: scalePoint(parsePoint(curvePoint.curveTo), frame),
    curveTo: scalePoint(parsePoint(curvePoint.curveFrom), frame),
    point: scalePoint(parsePoint(curvePoint.point), frame),
  };
}

export function encodeCurvePoint(
  curvePoint: DecodedCurvePoint,
  frame: Rect,
): Sketch.CurvePoint {
  return {
    ...curvePoint,
    hasCurveFrom: curvePoint.hasCurveTo,
    hasCurveTo: curvePoint.hasCurveFrom,
    curveFrom: stringifyPoint(unscalePoint(curvePoint.curveTo, frame)),
    curveTo: stringifyPoint(unscalePoint(curvePoint.curveFrom, frame)),
    point: stringifyPoint(unscalePoint(curvePoint.point, frame)),
  };
}

function curveTo(
  path: Path,
  current: DecodedCurvePoint,
  next: DecodedCurvePoint,
) {
  const curveTo = current.hasCurveTo ? current.curveTo : current.point;
  const curveFrom = next.hasCurveFrom ? next.curveFrom : next.point;
  const nextPoint = next.point;

  path.cubicTo(
    curveTo.x,
    curveTo.y,
    curveFrom.x,
    curveFrom.y,
    nextPoint.x,
    nextPoint.y,
  );
}

function lineTo(path: Path, next: DecodedCurvePoint) {
  const nextPoint = next.point;

  path.lineTo(nextPoint.x, nextPoint.y);
}

function curveOrLineTo(
  path: Path,
  current: DecodedCurvePoint,
  next: DecodedCurvePoint,
) {
  if (
    current.curveMode === Sketch.CurveMode.Straight &&
    next.curveMode === Sketch.CurveMode.Straight
  ) {
    lineTo(path, next);
  } else {
    curveTo(path, current, next);
  }
}

type CurveInfo = {
  radius: number;
  step: Point;
  drawSegment: boolean;
};

function stepToCurve(
  path: Path,
  current: DecodedCurvePoint,
  curveInfo: CurveInfo,
) {
  const currentPoint = current.point;

  const { radius, step } = curveInfo;

  if (radius > 0) {
    path.arcToTangent(
      currentPoint.x,
      currentPoint.y,
      currentPoint.x + step.x,
      currentPoint.y + step.y,
      radius,
    );
  } else {
    path.lineTo(currentPoint.x + step.x, currentPoint.y + step.y);
  }
}

export function path(
  CanvasKit: CanvasKit,
  points: Sketch.CurvePoint[],
  frame: Rect,
  isClosed: boolean,
): Path {
  const path = new CanvasKit.Path();

  if (points.length < 2) return path;

  const curvePoints = points.map((point) => decodeCurvePoint(point, frame));

  // Curves with only 2 points don't support corner radius, so we handle this
  // case specially
  if (curvePoints.length === 2) {
    const [current, next] = curvePoints;

    path.moveTo(current.point.x, current.point.y);

    curveOrLineTo(path, current, next);

    if (isClosed) {
      curveOrLineTo(path, next, current);

      path.close();
    }

    return path;
  }

  // Create an array of point tuples, [prev, current, next]
  const pairs = rotate(windowsOf(curvePoints, 3, true), -1);

  const curveInfoList: CurveInfo[] = pairs.map((pair) => {
    const [prev, current, next] = pair;

    const currentPoint = current.point;
    const nextPoint = next.point;
    const prevPoint = prev.point;

    const radius = getCornerRadius(
      prevPoint,
      currentPoint,
      nextPoint,
      current.cornerRadius,
    );

    const { step, drawSegment } = getStep(currentPoint, nextPoint, radius);

    return { radius, step, drawSegment };
  });

  zip(pairs, curveInfoList).forEach(([pair, curveInfo], index) => {
    const [, current, next] = pair;
    const isLast = index === pairs.length - 1;

    const currentPoint = current.point;
    const nextPoint = next.point;

    if (index === 0) {
      const { step } = curveInfo;

      path.moveTo(currentPoint.x + step.x, currentPoint.y + step.y);

      if (next.curveMode !== Sketch.CurveMode.Straight) {
        curveTo(path, current, next);
      }
    } else {
      if (
        // TODO: Handle one straight point with radius and one curve
        current.curveMode === Sketch.CurveMode.Straight &&
        next.curveMode === Sketch.CurveMode.Straight
      ) {
        stepToCurve(path, current, curveInfo);

        const { step, drawSegment } = curveInfo;

        if (drawSegment && (isClosed || !isLast)) {
          path.lineTo(nextPoint.x - step.x, nextPoint.y - step.y);
        }
      } else if (isClosed || !isLast) {
        curveTo(path, current, next);
      }
    }

    if (isClosed && isLast) {
      const [, current, next] = pairs[0];

      // If we have a corner radius, draw the final curve
      if (
        current.curveMode === Sketch.CurveMode.Straight &&
        next.curveMode === Sketch.CurveMode.Straight
      ) {
        stepToCurve(path, current, curveInfoList[0]);
      }

      // We need the path to be closed for line caps to draw correctly
      path.close();
    }
  });

  return path;
}

const getBorderPositionPathOp = (
  CanvasKit: CanvasKit,
  position: Sketch.BorderPosition,
): PathOp | undefined => {
  switch (position) {
    case Sketch.BorderPosition.Center:
      return;
    case Sketch.BorderPosition.Outside:
      return CanvasKit.PathOp.Difference;
    case Sketch.BorderPosition.Inside:
      return CanvasKit.PathOp.Intersect;
  }
};

export function getStrokedPath(
  CanvasKit: CanvasKit,
  path: Path,
  width: number,
  lineCap: Sketch.LineCapStyle,
  lineJoin: Sketch.LineJoinStyle,
  pathOp?: PathOp,
): Path {
  const copy = path.copy();

  // A 0-width path will never change the original when unioned
  if (width === 0 && pathOp === CanvasKit.PathOp.Union) return copy;

  if (
    !copy.stroke({
      width,
      cap: lineCapStyle(CanvasKit, lineCap),
      join: lineJoinStyle(CanvasKit, lineJoin),
    })
  ) {
    console.info('[getStrokedPath] Failed to stroke path');
    return copy;
  }

  if (pathOp !== undefined) {
    if (!copy.op(path, pathOp)) {
      console.info('[getStrokedPath] Failed to combine paths');
    }
  }

  return copy;
}

export function getStrokedBorderPath(
  CanvasKit: CanvasKit,
  path: Path,
  borderWidth: number,
  borderPosition: Sketch.BorderPosition,
  lineCap: Sketch.LineCapStyle,
  lineJoin: Sketch.LineJoinStyle,
) {
  switch (borderPosition) {
    case Sketch.BorderPosition.Center:
      break;
    case Sketch.BorderPosition.Outside:
    case Sketch.BorderPosition.Inside:
      borderWidth *= 2;
  }

  const pathOp = getBorderPositionPathOp(CanvasKit, borderPosition);

  return getStrokedPath(
    CanvasKit,
    path,
    borderWidth,
    lineCap,
    lineJoin,
    pathOp,
  );
}

export function unscaleCurvePoint(curvePoint: Sketch.CurvePoint, frame: Rect) {
  const unscale = (value: string) =>
    PointString.encode(unscalePoint(PointString.decode(value), frame));

  return {
    ...curvePoint,
    curveFrom: unscale(curvePoint.curveFrom),
    curveTo: unscale(curvePoint.curveTo),
    point: unscale(curvePoint.point),
  };
}

function pathCommandToSVGCommand(
  command: PathCommand,
): CommandWithoutQuadratics {
  switch (command[0]) {
    case PathCommandVerb.move: {
      const [, x, y] = command;
      return { type: 'move', to: { x, y } };
    }
    case PathCommandVerb.cubic: {
      const [, x1, y1, x2, y2, x3, y3] = command;
      return {
        type: 'cubicCurve',
        to: { x: x3, y: y3 },
        controlPoint1: { x: x1, y: y1 },
        controlPoint2: { x: x2, y: y2 },
      };
    }
    default:
      throw new Error('Path command not handled');
  }
}

export function pathToCurvePoints(
  path: Path,
  frame: Rect,
): Sketch.CurvePoint[] {
  const svgCommands = parsePathCmds(path.toCmds()).map(pathCommandToSVGCommand);

  // Assume a single path. `isClosed` should already be handled, before calling
  // this function, so we can ignore it here.
  const curvePoints = makePathsFromCommands(svgCommands)[0].points;

  return curvePoints.map((curvePoint) => unscaleCurvePoint(curvePoint, frame));
}

export function joinCurvePoints(
  segments: Sketch.CurvePoint[][],
  wrapsAround: boolean,
): Sketch.CurvePoint[] {
  const nonEmptySegments = segments.filter((segment) => segment.length > 0);

  if (nonEmptySegments.length === 0) return [];

  const [first, ...rest] = nonEmptySegments;

  const merged = rest.reduce((start, end) => {
    const lastStartPoint = start[start.length - 1];
    const firstEndPoint = end[0];

    return [
      ...start.slice(0, -1),
      SketchModel.curvePoint({
        curveMode: Sketch.CurveMode.Mirrored,
        hasCurveFrom: true,
        hasCurveTo: true,
        point: lastStartPoint.point,
        curveFrom: firstEndPoint.curveFrom,
        curveTo: lastStartPoint.curveTo,
      }),
      ...end.slice(1),
    ];
  }, first);

  if (wrapsAround) {
    const lastIndex = merged.length - 1;
    const last = merged[lastIndex];
    merged.splice(lastIndex, 1);
    merged[0].curveTo = last.curveTo;
  }

  return merged;
}

export function splitPath(path: Path, t: number): [Path, Path] {
  t = clamp(t, 0, 1);

  return [path.copy().trim(0, t, false)!, path.copy().trim(t, 1, false)!];
}
