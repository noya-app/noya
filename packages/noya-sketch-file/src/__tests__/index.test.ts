import fs from 'fs';
import path from 'path';
import { decode, encode } from '../index';

const sketchFile = fs.readFileSync(
  path.join(__dirname, 'fixtures/Rectangle.sketch'),
);

test('it should decode', async () => {
  const decoded = await decode(sketchFile);

  expect(decoded).toMatchSnapshot();
});

// We decode, re-encode, and re-decode the fixture, making sure both decoded
// versions match. If they do our `encode` probably works correctly.
test('it should encode', async () => {
  const decoded = await decode(sketchFile);
  const encoded = await encode(decoded);
  const decoded2 = await decode(Buffer.from(encoded));

  expect(decoded).toEqual(decoded2);
});
