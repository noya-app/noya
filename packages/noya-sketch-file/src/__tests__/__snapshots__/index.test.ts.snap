// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should decode 1`] = `
Object {
  "document": Object {
    "_class": "document",
    "assets": Object {
      "_class": "assetCollection",
      "colorAssets": Array [],
      "colors": Array [],
      "do_objectID": "142A5F95-0C29-4344-9427-220C0A9F3275",
      "exportPresets": Array [],
      "gradientAssets": Array [],
      "gradients": Array [],
      "imageCollection": Object {
        "_class": "imageCollection",
        "images": Object {},
      },
      "images": Array [],
    },
    "colorSpace": 0,
    "currentPageIndex": 0,
    "do_objectID": "1C727630-B463-43F8-9ACB-4061FE35251B",
    "documentState": Object {
      "_class": "documentState",
    },
    "fontReferences": Array [],
    "foreignLayerStyles": Array [],
    "foreignSwatches": Array [
      Object {
        "_class": "MSImmutableForeignSwatch",
        "do_objectID": "DC9835AA-09B9-4C28-B4B2-342F8889FBE0",
        "libraryID": "8902D83A-FDDA-4C49-9EF6-D2C58C9F71E9",
        "localSwatch": Object {
          "_class": "swatch",
          "do_objectID": "C9E211AA-59D2-48F4-870C-292C5D0AA032",
          "name": "5. Amazonite",
          "value": Object {
            "_class": "color",
            "alpha": 1,
            "blue": 0.7137254901960784,
            "green": 0.8431372549019608,
            "red": 0.2666666666666667,
          },
        },
        "remoteSwatchID": "EF7B5E12-0F72-42BB-8767-A6FFB04B9BEC",
        "sourceLibraryName": "Migrated Global Colors",
        "symbolPrivate": false,
      },
    ],
    "foreignSymbols": Array [],
    "foreignTextStyles": Array [],
    "layerStyles": Object {
      "_class": "sharedStyleContainer",
      "do_objectID": "7D3BE70A-942E-4E17-A817-000175E758D5",
      "objects": Array [],
    },
    "layerSymbols": Object {
      "_class": "symbolContainer",
      "do_objectID": "C928A506-B771-48BB-A684-872A1D0F36D1",
      "objects": Array [],
    },
    "layerTextStyles": Object {
      "_class": "sharedTextStyleContainer",
      "do_objectID": "BA21EEF8-8DB4-417A-97B5-4A2651FC8FBB",
      "objects": Array [],
    },
    "pages": Array [
      Object {
        "_class": "MSJSONFileReference",
        "_ref": "pages/D7CB52F1-C501-4F13-863B-A0D1B5FFA486",
        "_ref_class": "MSImmutablePage",
      },
    ],
    "sharedSwatches": Object {
      "_class": "swatchContainer",
      "do_objectID": "71A75562-8287-4001-B355-F08503EFDBF3",
      "objects": Array [],
    },
  },
  "images": Object {},
  "meta": Object {
    "app": "com.bohemiancoding.sketch3",
    "appVersion": "70.2",
    "autosaved": 0,
    "build": 109071,
    "commit": "f63936d75e8e7e0fa9213cfa45ffaf6ddd3e36ca",
    "compatibilityVersion": 99,
    "created": Object {
      "app": "com.bohemiancoding.sketch3",
      "appVersion": "70.2",
      "build": 109071,
      "commit": "f63936d75e8e7e0fa9213cfa45ffaf6ddd3e36ca",
      "compatibilityVersion": 99,
      "variant": "NONAPPSTORE",
      "version": 134,
    },
    "fonts": Array [],
    "pagesAndArtboards": Object {
      "D7CB52F1-C501-4F13-863B-A0D1B5FFA486": Object {
        "artboards": Object {},
        "name": "Page 1",
      },
    },
    "saveHistory": Array [
      "NONAPPSTORE.109071",
    ],
    "variant": "NONAPPSTORE",
    "version": 134,
  },
  "pages": Array [
    Object {
      "_class": "page",
      "booleanOperation": -1,
      "clippingMaskMode": 0,
      "do_objectID": "D7CB52F1-C501-4F13-863B-A0D1B5FFA486",
      "exportOptions": Object {
        "_class": "exportOptions",
        "exportFormats": Array [],
        "includedLayerIds": Array [],
        "layerOptions": 0,
        "shouldTrim": false,
      },
      "frame": Object {
        "_class": "rect",
        "constrainProportions": true,
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
      "groupLayout": Object {
        "_class": "MSImmutableFreeformGroupLayout",
      },
      "hasClickThrough": true,
      "hasClippingMask": false,
      "horizontalRulerData": Object {
        "_class": "rulerData",
        "base": 0,
        "guides": Array [],
      },
      "includeInCloudUpload": true,
      "isFixedToViewport": false,
      "isFlippedHorizontal": false,
      "isFlippedVertical": false,
      "isLocked": false,
      "isVisible": true,
      "layerListExpandedType": 0,
      "layers": Array [
        Object {
          "_class": "rectangle",
          "booleanOperation": -1,
          "clippingMaskMode": 0,
          "do_objectID": "A1C9FB6A-DD3E-4CB5-93EF-59FF90CCEBFA",
          "edited": false,
          "exportOptions": Object {
            "_class": "exportOptions",
            "exportFormats": Array [],
            "includedLayerIds": Array [],
            "layerOptions": 0,
            "shouldTrim": false,
          },
          "fixedRadius": 0,
          "frame": Object {
            "_class": "rect",
            "constrainProportions": false,
            "height": 226,
            "width": 266,
            "x": 356,
            "y": 228,
          },
          "hasClippingMask": false,
          "hasConvertedToNewRoundCorners": true,
          "isClosed": true,
          "isFixedToViewport": false,
          "isFlippedHorizontal": false,
          "isFlippedVertical": false,
          "isLocked": false,
          "isVisible": true,
          "layerListExpandedType": 0,
          "name": "Rectangle",
          "nameIsFixed": false,
          "needsConvertionToNewRoundCorners": false,
          "pointRadiusBehaviour": 1,
          "points": Array [
            Object {
              "_class": "curvePoint",
              "cornerRadius": 0,
              "curveFrom": "{0, 0}",
              "curveMode": 1,
              "curveTo": "{0, 0}",
              "hasCurveFrom": false,
              "hasCurveTo": false,
              "point": "{0, 0}",
            },
            Object {
              "_class": "curvePoint",
              "cornerRadius": 0,
              "curveFrom": "{1, 0}",
              "curveMode": 1,
              "curveTo": "{1, 0}",
              "hasCurveFrom": false,
              "hasCurveTo": false,
              "point": "{1, 0}",
            },
            Object {
              "_class": "curvePoint",
              "cornerRadius": 0,
              "curveFrom": "{1, 1}",
              "curveMode": 1,
              "curveTo": "{1, 1}",
              "hasCurveFrom": false,
              "hasCurveTo": false,
              "point": "{1, 1}",
            },
            Object {
              "_class": "curvePoint",
              "cornerRadius": 0,
              "curveFrom": "{0, 1}",
              "curveMode": 1,
              "curveTo": "{0, 1}",
              "hasCurveFrom": false,
              "hasCurveTo": false,
              "point": "{0, 1}",
            },
          ],
          "resizingConstraint": 63,
          "resizingType": 0,
          "rotation": 0,
          "shouldBreakMaskChain": false,
          "style": Object {
            "_class": "style",
            "blur": Object {
              "_class": "blur",
              "center": "{0.5, 0.5}",
              "isEnabled": false,
              "motionAngle": 0,
              "radius": 10,
              "saturation": 1,
              "type": 0,
            },
            "borderOptions": Object {
              "_class": "borderOptions",
              "dashPattern": Array [],
              "isEnabled": true,
              "lineCapStyle": 0,
              "lineJoinStyle": 0,
            },
            "borders": Array [],
            "colorControls": Object {
              "_class": "colorControls",
              "brightness": 0,
              "contrast": 1,
              "hue": 0,
              "isEnabled": false,
              "saturation": 1,
            },
            "contextSettings": Object {
              "_class": "graphicsContextSettings",
              "blendMode": 0,
              "opacity": 1,
            },
            "do_objectID": "FF9276E1-5D78-4B3B-9C8A-E483342A4751",
            "endMarkerType": 0,
            "fills": Array [
              Object {
                "_class": "fill",
                "color": Object {
                  "_class": "color",
                  "alpha": 1,
                  "blue": 1,
                  "green": 0.6895309893279358,
                  "red": 0.2805640243902439,
                },
                "contextSettings": Object {
                  "_class": "graphicsContextSettings",
                  "blendMode": 0,
                  "opacity": 1,
                },
                "fillType": 0,
                "gradient": Object {
                  "_class": "gradient",
                  "elipseLength": 0,
                  "from": "{0.5, 0}",
                  "gradientType": 0,
                  "stops": Array [
                    Object {
                      "_class": "gradientStop",
                      "color": Object {
                        "_class": "color",
                        "alpha": 1,
                        "blue": 1,
                        "green": 1,
                        "red": 1,
                      },
                      "position": 0,
                    },
                    Object {
                      "_class": "gradientStop",
                      "color": Object {
                        "_class": "color",
                        "alpha": 1,
                        "blue": 0,
                        "green": 0,
                        "red": 0,
                      },
                      "position": 1,
                    },
                  ],
                  "to": "{0.5, 1}",
                },
                "isEnabled": true,
                "noiseIndex": 0,
                "noiseIntensity": 0,
                "patternFillType": 1,
                "patternTileScale": 1,
              },
            ],
            "innerShadows": Array [],
            "miterLimit": 10,
            "shadows": Array [],
            "startMarkerType": 0,
            "windingRule": 1,
          },
        },
      ],
      "name": "Page 1",
      "nameIsFixed": false,
      "resizingConstraint": 63,
      "resizingType": 0,
      "rotation": 0,
      "shouldBreakMaskChain": false,
      "style": Object {
        "_class": "style",
        "blur": Object {
          "_class": "blur",
          "center": "{0.5, 0.5}",
          "isEnabled": false,
          "motionAngle": 0,
          "radius": 10,
          "saturation": 1,
          "type": 0,
        },
        "borderOptions": Object {
          "_class": "borderOptions",
          "dashPattern": Array [],
          "isEnabled": true,
          "lineCapStyle": 0,
          "lineJoinStyle": 0,
        },
        "borders": Array [],
        "colorControls": Object {
          "_class": "colorControls",
          "brightness": 0,
          "contrast": 1,
          "hue": 0,
          "isEnabled": false,
          "saturation": 1,
        },
        "contextSettings": Object {
          "_class": "graphicsContextSettings",
          "blendMode": 0,
          "opacity": 1,
        },
        "do_objectID": "290F12BF-153C-493E-A31F-A4F798A2BEC8",
        "endMarkerType": 0,
        "fills": Array [],
        "innerShadows": Array [],
        "miterLimit": 10,
        "shadows": Array [],
        "startMarkerType": 0,
        "windingRule": 1,
      },
      "verticalRulerData": Object {
        "_class": "rulerData",
        "base": 0,
        "guides": Array [],
      },
    },
  ],
  "user": Object {
    "D7CB52F1-C501-4F13-863B-A0D1B5FFA486": Object {
      "scrollOrigin": "{0, 0}",
      "zoomValue": 1,
    },
    "document": Object {
      "componentSidebarTreeStructure": 0,
      "pageListCollapsed": 0,
      "pageListHeight": 182.5,
    },
  },
}
`;
